### PR TITLE
Allow rules to be tailored on profiles#create API

### PR DESCRIPTION
**Pull request by akofink**
POST body containing rules relationships:

```js
{
	"data": {
		"attributes": {
			"parent_profile_id": "1007b72e-7f0f-486f-911a-5c392c724faa",
			"name": "My tailored profile",
			"compliance_threshold": 95.0
		},
		"relationships": {
			"rules": {
				"data": [
					{
						"id": "cc9afa66-3536-4d2e-bc8e-10111d13ec50",
						"type": "rule"
					},
					{
						"id": "06a19f0e-5c7a-4d54-bc66-e932a96bf954",
						"type": "rule"
					},
					{
						"id": "fcd8ee29-5af4-47bf-9e29-2de12f20fe9a",
						"type": "rule"
					}
				]
			}
		}
	}
}

# Response:

HTTP/1.1 201 Created
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: X-Requested-With,content-type
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, PATCH, DELETE
Access-Control-Allow-Origin: *
cache-control: max-age=0, private, must-revalidate
connection: close
content-type: application/vnd.api+json
date: Tue, 23 Jun 2020 15:42:29 GMT
etag: W/"62bba6ee45343fbc2bbe364beb645a35"
transfer-encoding: chunked
vary: Origin
x-request-id: 31f18b51-105c-48b6-8844-880f335f5162
x-runtime: 0.275060

{
    "data": {
        "attributes": {
            "business_objective": null,
            "canonical": false,
            "compliance_threshold": 95.0,
            "compliant_host_count": 0,
            "description": "Ensures PCI-DSS v3.2.1 related security configuration settings are applied.",
            "external": false,
            "name": "My tailored profile",
            "os_major_version": "8",
            "parent_profile_id": "1007b72e-7f0f-486f-911a-5c392c724faa",
            "parent_profile_ref_id": "xccdf_org.ssgproject.content_profile_pci-dss",
            "ref_id": "xccdf_org.ssgproject.content_profile_pci-dss",
            "score": 0,
            "tailored": true,
            "total_host_count": 0
        },
        "id": "679133ea-bbe0-47ed-ae36-4093d4402b03",
        "relationships": {
            "account": {
                "data": {
                    "id": "8c926d2a-17d6-4398-8895-2652971113c1",
                    "type": "account"
                }
            },
            "benchmark": {
                "data": {
                    "id": "c0e2ce89-fb9a-4e29-a51b-23f2398cde83",
                    "type": "benchmark"
                }
            },
            "hosts": {
                "data": []
            },
            "parent_profile": {
                "data": {
                    "id": "1007b72e-7f0f-486f-911a-5c392c724faa",
                    "type": "profile"
                }
            },
            "rules": {
                "data": [
                    {
                        "id": "06a19f0e-5c7a-4d54-bc66-e932a96bf954",
                        "type": "rule"
                    },
                    {
                        "id": "cc9afa66-3536-4d2e-bc8e-10111d13ec50",
                        "type": "rule"
                    },
                    {
                        "id": "fcd8ee29-5af4-47bf-9e29-2de12f20fe9a",
                        "type": "rule"
                    }
                ]
            },
            "test_results": {
                "data": []
            }
        },
        "type": "profile"
    }
}
```

Note that the shape of the request and response are the same, other than unspecified attributes or relationships in the request.

On create, if no valid rules are specified, the parent profile rules are added to the new profile. Valid rules are those that exist in the parent profile's benchmark only.

Signed-off-by: Andrew Kofink <akofink@redhat.com>

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -21,12 +20,6 @@ def pundit_scope
     Pundit.policy_scope(current_user, resource)
   end
 
-  def resource_params
-    params.permit(data: ParamType.map(attributes: ParamType.map))
-    params.require(:data).require(:attributes)
-          .permit(serializer.attributes_to_serialize.keys)
-  end
-
```
Comment (akofink) [inline app/controllers/application_controller.rb:29]: Moved to `Parameters` concern

File: app/controllers/concerns/parameters.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# Reusable parameter checking for all controllers
+module Parameters
+  extend ActiveSupport::Concern
+
+  ParamType = ActionController::Parameters
+
+  RELATIONSHIP_TYPE = ParamType.map(
+    id: ParamType.string,
+    type: ParamType.string
+  )
```
Comment (akofink) [inline app/controllers/concerns/parameters.rb:12]: Type validations thus far are for request shape only (i.e. data, attributes, relationships are all `ParamType.map`) and not individual attribute types with the exception of type/id here which holds for any relationship.

File: app/controllers/v1/benchmarks_controller.rb [hunk 3]
```diff
@@ -30,5 +28,9 @@ def scope
     def resource
       Xccdf::Benchmark
     end
+
+    def serializer
+      BenchmarkSerializer
+    end
```
Comment (akofink) [inline app/controllers/v1/benchmarks_controller.rb:34]: Changes in this file are to allow users to see which rules exist on a benchmark, via its relationships and optional includes, so they can select rules for tailoring a profile.

File: app/controllers/v1/profiles_controller.rb [hunk 4]
```diff
@@ -27,6 +27,7 @@ def create
       profile = Profile.new(profile_create_attributes).fill_from_parent
 
       if profile.save
+        profile.update_rules(ids: new_rule_ids)
```
Comment (akofink) [inline app/controllers/v1/profiles_controller.rb:30]: This no longer happens on save automatically.

File: app/models/profile.rb [hunk 5]
```diff
@@ -84,31 +83,37 @@ def in_account(account)
...
+  def update_rules(ids: nil, ref_ids: nil)
+    ids_to_add = rule_ids_to_add(ids, ref_ids)
+
+    ProfileRule.where(rule_id: rule_ids - ids_to_add, profile_id: id)
+               .destroy_all
+
+    ProfileRule.import!(ids_to_add.map do |rule|
+      ProfileRule.new(profile_id: id, rule_id: rule.id)
+    end)
+  end
```
Comment (akofink) [inline app/models/profile.rb:100]: Renamed from `add_rules`. This method now intelligently updates the rules on the profile and will remove rules that aren't specified. Also, `ignore: true` was removed from the import, as we don't expect duplicates or any other errors anymore.

File: config/initializers/wrap_parameters.rb [hunk 6]
```diff
@@ -7,7 +7,7 @@
 
 # Enable parameter wrapping for JSON. You can disable this by setting :format to an empty array.
 ActiveSupport.on_load(:action_controller) do
-  wrap_parameters format: [:json]
+  wrap_parameters format: []
```
Comment (akofink) [inline config/initializers/wrap_parameters.rb:10]: Before this change, we would always get an extra param sent to the controller named after the resource (i.e. `params[:profile]`). That doesn't fit with our JSON spec where everything is wrapped in `params[:data]`.

File: app/serializers/benchmark_serializer.rb [hunk 7]
```diff
@@ -4,4 +4,6 @@
 class BenchmarkSerializer
   include FastJsonapi::ObjectSerializer
   attributes :ref_id, :title, :version, :description
+  has_many :rules
+  has_many :profiles
```
[the remark this answers is not in the document]
Comment (akofink) [reply app/serializers/benchmark_serializer.rb:8]: This should be limited by the pundit policy, but it's a good thing for us to double check, for sure. It makes sense for a user to want to see all profiles, even non-canonical, associated to a specific benchmark in their account, but this API will most likely be used for showing canonical profiles, so perhaps we should add a default `canonical = true` search to the benchmarks controller. What do you think?

File: app/models/profile.rb (hunk 5 above)
Comment (dLobatog) [reply app/models/profile.rb:100]: This is actually pretty useful, I am working on making the frontend Edit Policy button bring up the wizard to allow re-tailoring and this comes in handy

File: spec/integration/profiles_spec.rb [hunk 8]
```diff
@@ -97,6 +100,14 @@
               'of these checks should pass.',
               compliance_threshold: 95.0,
               business_objective: 'APAC Expansion'
+            },
+            relationships: {
```
[the remark this answers is not in the document]
Comment (akofink) [reply spec/integration/profiles_spec.rb:104]: Added. :+1:

File: app/serializers/benchmark_serializer.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (akofink) [reply app/serializers/benchmark_serializer.rb]: Nice catch, @dLobatog - this was indeed showing profiles from other user accounts. Now I've limited it by pundit policy. We'll have to be careful in the future to do this for relationships in serializers.

**Comment by akofink** [comment]
Updated! I think all comments so far have been addressed.

File: app/serializers/benchmark_serializer.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (dLobatog) [reply app/serializers/benchmark_serializer.rb]: Definitely, top-level objects are properly filtered via pundit easily but serialized relations have to be also filtered

**Pull request merged**